subprojects/docs/src/docs/userguide/tasks.xml

```diff
@@ -290,52 +290,673 @@
+                        <para>For this example, we’re using a custom class to represent the model data.
+                            However, it does not implement <literal>Serializable</literal>, so we can’t
+                            use the <literal>@Input</literal> annotation. That’s not a problem as the
+                            properties within <literal>TemplateData</literal> - a string and a hash map
+                            with serializable type parameters - are serializable and can be annotated with
+                            <literal>@Input</literal>. We use <literal>@Nested</literal> on
+                            <literal>templateData</literal> to let Gradle know that this is a value with
+                            nested input properties.
+                       </para>
+                    </listitem>
+                    <listitem><para><literal>outputDir</literal></para>
+                        <para>The directory where the generated files go. As with input files, there are
+                            several annotations for output files and directories. A property representing
+                            a single directory requires <literal>@OutputDirectory</literal>. You’ll learn
+                            about the others soon.
+                        </para>
+                    </listitem>
+                </itemizedlist>
+
+                <para>These annotated properties mean that Gradle will skip the task if none of the source
+                    files, template engine, model data or generated files have changed since the previous
+                    time Gradle executed the task. This will often save a significant amount of time. You
+                    can learn how Gradle detects <xref linkend="sec:how_does_inc_build_work">changes later</xref>.
+                </para>
+
+                <para>This example is particularly interesting because it works with collections of source
+                    files. What happens if only one source file changes? Does the task process all the
+                    source files again or just the modified one? That depends on the task implementation.
+                    If the latter, then the task itself is incremental, but that’s a different feature to
+                    the one we’re discussing here. Gradle does help task implementers with this via its
+                    <xref linkend="incremental_tasks">incremental task inputs</xref> feature.
+                </para>
+
+                <para>Now that you have seen some of the input and output annotations in practice, let’s
+                    take a look at all the annotations available to you and when you should use them. The
+                    table below lists the available annotations and the corresponding property type you
+                    can use with each one.
+                </para>
+                <table>
+                    <title>Incremental build annotations</title>
```

Comment:
That was deliberate because it's specific to validation. That said, since I've included `@SkipWhenEmpty`, perhaps it makes sense to include `@Optional` as well and just link to the later section.